### PR TITLE
make afl-system-config available as system-config

### PR DESCRIPTION
**Pull request by vanhauser-thc**

### CODE REVIEW

**Comment by vanhauser-thc** [comment]
I am not a big rust guy :)
now that you point it out, "sudo cargo run afl system-config" could result in a lot of files being created by root that will make future compilations fails if done without sudo.

IMHO  `cargo run afl system-config` is important to have, but it should be done different.
how about running it "sudo afl-system-config" and bail if that fails with an error that helps what should be done?

**Comment by vanhauser-thc** [comment]
how about this?

**Comment by smoelius** [comment]
> how about running it "sudo afl-system-config" and bail if that fails with an error that helps what should be done?

Just to be sure we're on the same page, you're suggesting that the `afl-system-config` subcommand be "special":
- All of the other subcommands continue to run an afl++ binary or Cargo subcommand directly.
- But rather than run `<path-to-afl-system-config>`, the `afl-system-config` subcommand runs `sudo <path-to-afl-system-config>`.

Is that right?

**Comment by vanhauser-thc** [comment]
Yes exactly

File: afl/src/bin/cargo-afl.rs [hunk 1]
```diff
@@ -224,11 +244,27 @@ where
     I: IntoIterator<Item = S>,
     S: AsRef<OsStr>,
 {
-    let cmd_path = common::afl_dir(None).join("bin").join(tool);
-    let mut cmd = Command::new(cmd_path);
-    cmd.args(args);
+    let cmd_path: PathBuf;
+    let mut cmd;
+    if tool.to_string() == "sudo".to_string() {
```
Comment (smoelius) [inline afl/src/bin/cargo-afl.rs:249]: Could we structure the code involving this `if` block as follows?
```rust
let mut cmd = if tool == "afl-system-config" {
    // sudo stuff
} else {
   ...
};
cmd.args(args);
// Rest as before.
```
Also, could we please add `--reset-timestamp` to the sudo invocation? This way, if the user is surprised by the sudo invocation, they have a chance to cancel it.

File: afl/src/bin/cargo-afl.rs [hunk 2]
```diff
@@ -224,11 +244,27 @@ where
...
+        let arguments = vec.into_iter();
+        cmd.args(arguments);
+    } else {
+        cmd_path = common::afl_dir(None).join("bin").join(tool);
+        cmd = Command::new(cmd_path);
+        cmd.args(args);
+    }
+
     let status = cmd.status().unwrap();
-    #[cfg(target_os = "macos")]
```
Comment (smoelius) [inline afl/src/bin/cargo-afl.rs:231]: Why did you remove this? Can the `shmget` error message appear on systems besides macOS? (That's the only place I've known it to occur.)

But while we are at it, could you please update the help message so that it refers to `cargo afl system-config`?

File: afl/src/bin/cargo-afl.rs [hunk 3]
```diff
@@ -382,7 +418,15 @@ mod tests {
...
-        "analyze", "cmin", "fuzz", "gotcpu", "plot", "showmap", "tmin", "whatsup",
+        "analyze",
+        "cmin",
+        "fuzz",
+        "gotcpu",
+        "system-config",
+        "plot",
+        "showmap",
+        "tmin",
+        "whatsup",
```
Comment (smoelius) [inline afl/src/bin/cargo-afl.rs:429]: How did you choose this order? In the absence of some other reason, I would prefer to default to alphabetical order, like before.

Also, could you please ensure that the order in which the subcommands are listed above matches how they are listed here?

**Review by smoelius — CHANGES_REQUESTED** [review]
Thank you again for this PR. I agree that having a `system-config` subcommand will be convenient.

File: afl/src/bin/cargo-afl.rs [hunk 4]
```diff
@@ -164,6 +171,19 @@ fn clap_app() -> clap::Command {
...
+                        .allow_hyphen_values(true)
+                        .disable_help_subcommand(true)
+                        .disable_help_flag(true)
+                        .disable_version_flag(true)
+                        .arg(
+                            Arg::new("sudo args")
+                                .value_parser(value_parser!(OsString))
+                                .num_args(0..),
+                        ),
+                )
```
[the remark this answers is not in the document]
Comment (smoelius) [reply afl/src/bin/cargo-afl.rs:186]: Oh, but please keep the `(beware, called with sudo!)` message.

File: afl/src/bin/cargo-afl.rs (hunk 2 above)
Comment (vanhauser-thc) [reply afl/src/bin/cargo-afl.rs:231]: I removed it because shm_open can fail on Linux too. Or any other system config error that afl-fuzz complains about - which can all be fixed with afl-system-config.

**Comment by vanhauser-thc** [comment]
I implemented all changes with the exception to make the sudo call stuff prettier/streamlined. this is beyond my limited rust skills ... :(

**Comment by smoelius** [comment]
I'm trying to understand the CI failures, which has caused me to find this explanation of `LaunchAgents` vs. `LaunchDaemons`: https://ss64.com/osx/launchctl.html

Furthermore, I see [these lines](https://github.com/AFLplusplus/AFLplusplus/blob/f596a297c4de6a5e1a6fb9fbb3b4e18124a24f58/afl-system-config#L117-L118) in `afl-system-config` (note the use of `sudo` on the second line):
```sh
    launchctl unload -w ${SL}/LaunchAgents/${PL}.plist >/dev/null 2>&1
    sudo launchctl unload -w ${SL}/LaunchDaemons/${PL}.Root.plist >/dev/null 2>&1
```
That is, for Mac, it looks like `afl-system-config` is expected to be run not under `sudo`.

Am I misinterpreting the above lines?

**Comment by smoelius** [comment]
FWIW, here is a slight reworking of `afl-system-config` that expects to be called with `sudo`: https://github.com/AFLplusplus/AFLplusplus/compare/stable...smoelius:AFLplusplus:stable

With this variant, [CI passes](https://github.com/rust-fuzz/afl.rs/actions/runs/5912174826).

How do you think we should proceed?

**Comment by vanhauser-thc** [comment]
> That is, for Mac, it looks like afl-system-config is expected to be run not under sudo.

no it needs to run under root/run with sudo. I do not know why there is an extra sudo in there :-) I did not code that part. but it works :) 

I merged your changes to afl-system-config, I booted up my macos mini I never use and it didnt work before, with your patch it does.

so this will need a new afl++ release or are you fine to pin AFL++ to the current stable branch?

File: afl/src/bin/cargo-afl.rs [hunk 5]
```diff
@@ -274,7 +262,7 @@ If you see an error message like `shmget() failed` above, try running the follow
 
     cargo-afl afl system-config
 
-Note: You might be prompted to enter your password as root privileges are required and hence sudo is run within this command."
+Note: You will be prompted to enter your password as root privileges are required and hence sudo is run within this command."
```
Comment (vanhauser-thc) [inline afl/src/bin/cargo-afl.rs:267]: fyi I put "might" because if the sudo user is configured with NOPASSWD or already used sudo in the shell then there will be no password prompt.

File: afl/src/bin/cargo-afl.rs [hunk 6]
```diff
@@ -224,21 +243,28 @@ where
...
         eprintln!(
             "
 If you see an error message like `shmget() failed` above, try running the following command:
 
-    sudo {}
+    cargo-afl afl system-config
 
-Note: You will be prompted to enter your password.",
-            sudo_cmd_path.display()
+Note: You will be prompted to enter your password as root privileges are required and hence sudo is run within this command."
```
Comment (smoelius) [inline afl/src/bin/cargo-afl.rs:267]:
```suggestion
Note: You might be prompted to enter your password as root privileges are required and hence sudo is run within this command."
```

FWIW, I added `--reset-timestamp` to the sudo invocation in attempt to address the "already used sudo" problem.

But I see your point about NOPASSWD.

File: README.md [hunk 7]
```diff
@@ -43,6 +43,11 @@ Screen recording of AFL running on Rust code.
 
 ## Hints
 
+Before starting to fuzz, you should reconfigure your system for optimal
+performance and better crash detection. This can be done with `cargo-afl afl system-config`.
+But this subcommand requires root, so it uses sudo internally. Hence, you will need to enter
```
Comment (smoelius) [inline README.md:48]:
```suggestion
But this subcommand requires root, so it uses sudo internally. Hence, you might need to enter
```
To make this consistent with the help message below. (Sorry for going back and forth.)

**Comment by smoelius** [comment]
> I merged your changes to afl-system-config, I booted up my macos mini I never use and it didnt work before, with your patch it does.

I'm glad to hear that!

> so this will need a new afl++ release or are you fine to pin AFL++ to the current stable branch?

I'm fine to pin to the current stable branch for now.

I would like to treat this new subcommand as a breaking change. There are other changes I want to include in the next breaking-change release (e..g., [this](https://github.com/rust-fuzz/afl.rs/issues/335) and [this](https://github.com/rust-fuzz/afl.rs/issues/369)), and it will take some time for me to prepare them.

If a new AFL++ release is not out before then, we can revisit.

Can I impose on you to commit the last two suggestions and update the AFL++ submodule? After that, I think this should be mergeable.

**Comment by vanhauser-thc** [comment]
Made the last changes, update the AFL++ commit, CI is now green.
IHMO it is not a breaking change. It adds a new command, so everyone can just work as before and ignore it :)
but up to you obviously.

**Review by smoelius — APPROVED** [review]

**Comment by smoelius** [comment]
> IHMO it is not a breaking change. It adds a new command, so everyone can just work as before and ignore it :)

You're right, of course. My reason for wanting to consider the change breaking is that if (hypothetically) a user objects to `cargo-afl` calling sudo, they can stick with version 0.13.

Thanks a lot for all of your hard work on this.
